chain/errors.go

```diff
@@ -0,0 +1,495 @@
+	ErrMissingInputs
+
+	// ErrOversizeTx is returned when a transaction is too large.
+	ErrOversizeTx
+
+	// ErrCoinbaseTx is returned when the transaction is coinbase tx.
+	ErrCoinbaseTx
+
+	// ErrNonStandardVersion is returned when the transactions are not
+	// standard - a version currently non-standard.
+	ErrNonStandardVersion
+
+	// ErrNonStandardScript is returned when the transactions are not
+	// standard - non-standard script.
+	ErrNonStandardScript
+
+	// ErrBareMultiSig is returned when the transactions are not standard -
+	// bare multisig script (2-of-3).
+	ErrBareMultiSig
+
+	// ErrScriptSigNotPushOnly is returned when the transactions are not
+	// standard - not-pushonly scriptSig.
+	ErrScriptSigNotPushOnly
+
+	// ErrScriptSigSize is returned when the transactions are not standard
+	// - too large scriptSig (>1650 bytes).
+	ErrScriptSigSize
+
+	// ErrTxTooLarge is returned when the transactions are not standard -
+	// too large tx size.
+	ErrTxTooLarge
+
+	// ErrDust is returned when the transactions are not standard - output
+	// too small.
+	ErrDust
+
+	// ErrMultiOpReturn is returned when the transactions are not standard
+	// - muiltiple OP_RETURNs.
+	ErrMultiOpReturn
+
+	// ErrNonFinal is returned when spending a timelocked transaction that
+	// hasn't expired yet.
+	ErrNonFinal
+
+	// ErrNonBIP68Final is returned when a transaction that is locked by
+	// BIP68 sequence logic and not expired yet.
+	ErrNonBIP68Final
+
+	// ErrSameNonWitnessData is returned when another tx with the same
+	// non-witness data is already in the mempool. For instance, these two
+	// txns share the same `txid` but different `wtxid`.
+	ErrSameNonWitnessData
+
+	// ErrNonMandatoryScriptVerifyFlag is returned when passing a raw tx to
+	// `testmempoolaccept`, which gives the error followed by (Witness
+	// program hash mismatch).
+	ErrNonMandatoryScriptVerifyFlag
+
+	// errSentinel is used to indicate the end of the error list. This
+	// should always be the last error code.
+	errSentinel
+)
+
+// Error implements the error interface. It returns the error message defined
+// in `bitcoind`.
+
+// Some of the dashes used in the original error string is removed, e.g.
+// "missing-inputs" is now "missing inputs". This is ok since we will normalize
+// the errors before matching.
+//
+// references:
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/rpc_rawtransaction.py#L342
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/data/invalid_txs.py
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_accept.py
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_accept_wtxid.py
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_dust.py
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_limit.py
+// - https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp
+func (r RPCErr) Error() string {
```

Comment:
Q: I wonder why we don't call this error in particular bitcoind error, because it reflects the errors of this particular backend ? Calling it rpcerr, I always think this is a generalized error ?

Reply:
yeah originally it was named `BitcoindRPCErr`, but in reality this functions as the generalized error - when we use it in `lnd`, we don't check for backends, but just check against, say, `chain.ErrMempoolConflict`.

Ideally, we can define another set of errors and put them in a map similar to `BtcdErrMap`, keyed by the error string returned from `bitcoind` and valued by the `RPCErr`, which brings more symmetry. However, unless we have a very different error string returned from the `chain`, there's no point in doing so as in the current design, we always use the error string defined in `bitcoind`. In other words, this `Error()` needs to return something that's human-readable, and we decide to use the error string from `bitcoind`.